ambry-commons/src/main/java/com.github.ambry.commons/HelixNotifier.java

```diff
@@ -0,0 +1,209 @@
+        storeFactory.getHelixPropertyStore(zkClientConnectString, zkClientSessionTimeoutMs, zkClientConnectionTimeoutMs,
+            rootPath, subscribedPaths);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Return {@code true} does not guarantee all the {@link TopicListener} will receive the message. It just indicates
+   * the message has been successfully dent out.
+   */
+  @Override
+  public boolean sendMessage(String topic, String message) {
+    if (topic == null) {
+      throw new IllegalArgumentException("path to listen cannot be null");
+    }
+    if (message == null) {
+      throw new IllegalArgumentException("message cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    ZNRecord record = new ZNRecord(topicPath);
+    record.setSimpleField(MESSAGE_KEY, message);
+    return helixStore.set(topicPath, record, AccessOption.PERSISTENT);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void subscribe(String topic, final TopicListener<String> topicListener) {
+    if (topic == null) {
+      throw new IllegalArgumentException("path to listen cannot be null");
+    }
+    if (topicListener == null) {
+      throw new IllegalArgumentException("listener cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    HelixPropertyListener helixListener = new HelixPropertyListener() {
+      @Override
+      public void onDataChange(String path) {
+        logger.trace("Message is changed for topic {} at path {}", topic, path);
+        sendMessage(topicListener, topic, path);
+      }
+
+      @Override
+      public void onDataCreate(String path) {
+        logger.trace("Message is created for topic {} at path {}", topic, path);
+        sendMessage(topicListener, topic, path);
+      }
+
+      @Override
+      public void onDataDelete(String path) {
+        // for now, this is a no-op when a ZNRecord for a topic is deleted, since no
+        // topic deletion is currently supported.
+        logger.trace("Message is deleted for topic {} at path {}", topic, path);
+      }
+    };
+    topicListenerToHelixListenerMap.put(topicListener, helixListener);
+    List<TopicListener> topicListeners = topicToListenerMap.get(topic);
+    if (topicListeners == null) {
+      topicListeners = new ArrayList<>();
+    }
+    topicListeners.add(topicListener);
+    topicToListenerMap.put(topic, topicListeners);
+    helixStore.subscribe(topicPath, helixListener);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void unsubscribe(String topic, TopicListener<String> topicListener) {
+    if (topic == null) {
+      throw new IllegalArgumentException("topic cannot be null");
+    }
+    if (topicListener == null) {
+      throw new IllegalArgumentException("topicListener cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    HelixPropertyListener helixListener = topicListenerToHelixListenerMap.remove(topicListener);
+    helixStore.unsubscribe(topicPath, helixListener);
```

Comment:
If a valid but already-unsubscribed `helixListener` is passed to a `HelixPropertyStore`, it will be silent per helix behavior. So I have kept the behavior the same. Actually this is a corner case that has not been covered by unit test, so I added the case in the unit test.